packages/plugin-patch/sources/PatchFetcher.ts

```diff
@@ -44,56 +44,87 @@ export class PatchFetcher implements Fetcher {
     const patchFiles = await patchUtils.loadPatchFiles(parentLocator, patchPaths, opts);
 
     const tmpDir = await xfs.mktempPromise();
-    const tmpFile = ppath.join(tmpDir, `patched.zip` as Filename);
+    const currentFile = ppath.join(tmpDir, `current.zip` as Filename);
 
     const sourceFetch = await opts.fetcher.fetch(sourceLocator, opts);
     const prefixPath = structUtils.getIdentVendorPath(locator);
 
     const libzip = await getLibzipPromise();
 
-    const patchedPackage = new ZipFS(tmpFile, {
+    // First we create a copy of the package that we'll be free to mutate
+    const initialCopy = new ZipFS(currentFile, {
       libzip,
       create: true,
       level: opts.project.configuration.get(`compressionLevel`),
     });
 
-    await patchedPackage.mkdirpPromise(prefixPath);
+    await initialCopy.mkdirpPromise(prefixPath);
 
     await miscUtils.releaseAfterUseAsync(async () => {
-      await patchedPackage.copyPromise(prefixPath, sourceFetch.prefixPath, {baseFs: sourceFetch.packageFs, stableSort: true});
+      await initialCopy.copyPromise(prefixPath, sourceFetch.prefixPath, {baseFs: sourceFetch.packageFs, stableSort: true});
     }, sourceFetch.releaseFs);
 
-    const patchFs = new CwdFS(ppath.resolve(PortablePath.root, prefixPath), {baseFs: patchedPackage});
-
-    for (const patchFile of patchFiles) {
-      if (patchFile !== null) {
-        try {
-          await patchUtils.applyPatchFile(patchUtils.parsePatchFile(patchFile), {
-            baseFs: patchFs,
-            version: sourceVersion,
-          });
-        } catch (err) {
-          if (!(err instanceof UnmatchedHunkError))
-            throw err;
-
-          const enableInlineHunks = opts.project.configuration.get(`enableInlineHunks`);
-          const suggestion = !enableInlineHunks
-            ? ` (set enableInlineHunks for details)`
-            : ``;
-
-          throw new ReportError(MessageName.PATCH_HUNK_FAILED, err.message + suggestion, report => {
-            if (!enableInlineHunks)
-              return;
-
-            reportHunk(err.hunk, {
-              configuration: opts.project.configuration,
-              report,
-            });
+    initialCopy.saveAndClose();
+
+    for (const {source, optional} of patchFiles) {
+      if (source === null)
+        continue;
+
+      // Then for each patchfile, we open this copy anew, and try to apply the
+      // changeset. We need to open it for each patchfile (rather than only a
+      // single time) because it lets us easily rollback when hitting errors
+      // on optional patches (we just need to call `discardAndClose`).
+      const patchedPackage = new ZipFS(currentFile, {
```

Comment:
Could you change this part and the `initialCopy` to be in-memory, then write the resulting buffer to disk for the returned ZipFS instance?

Reply:
I don't think I can, since we need to take snapshots in-between each patch. If I was to closeAndSave a memory zipFs, then we'd to already have in memory another one representing the last snapshot; this would be difficult unless we clone the memory zipFs somehow, which I don't know how to do efficiently.

Reply:
@arcanis Wouldn't this work? https://github.com/yarnpkg/berry/compare/mael/optional-patches...paul/patch-example

Reply:
Oh I forgot `getBufferAndClose` 🤔 Still, it would require to keep the whole archive in memory, twice per package. I'm a bit worried it would be heavy (by contrast, by closing / opening, it only needs the archive listing)... given that the general case is to apply a single patch, I'd tend to first try the fs approach and switch only if we see a noticeable impact.

Reply:
Can we at least profile both approaches and compare the time spent and the memory usage to make sure that we don't introduce a significant performance regression? Since compression is the slowest part of the fetch step, I'd prefer to avoid having to compress the archive for each patch file. I've also just realized that my changes still aren't optimal because they don't prevent compression (I have to pass `level: 0` for that to happen).

Reply:
Sure, will check - I however expect the compression level to affect new writes, not the existing ones. With that in mind we can disable compression, since it would require to go back and compress the affected files after all the patches have been applied - far too complex for a fringe use case (multiple patches on the same package).

Reply:
For the record: 22s install time for TS code cache, vs 15s before.